### PR TITLE
Add basic version of torch dataset

### PR DESCRIPTION
**Pull request by Flova**
# Summary

## Checklist

- [ ] Run `colcon build`
- [ ] Write documentation
- [ ] Create issues for future work
- [x] Test on your machine
- [ ] Test on the robot
- [x] This PR is on our `Software` project board

### CODE REVIEW

File: ddlitlab2024/ml/training/train.py [hunk 1]
```diff
@@ -1,130 +1,129 @@
...
     ).to(device)
 
     # Add normalization parameters to the model
-    model.mean = torch.tensor(means.values).to(device)
-    model.std = torch.tensor(stds.values).to(device)
+    model.mean = normalizer.mean
+    model.std = normalizer.std
 
+    # Utilize an Exponential Moving Average (EMA) for the model to smooth out the training process
     ema = EMA(model, beta=0.9999)
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:86]: For example here I would appreciate a comment, why we choose specifically this beta especially for people like me who have no experience with machine learning.
Rather than comments like:
```python
# Creating test
test = lib.test(...)
```

File: ddlitlab2024/ml/training/train.py [hunk 2]
```diff
@@ -1,130 +1,129 @@
...
+            joint_targets = batch["joint_command"]
 
-            # Split the data into past actions and noisy action predictions
-            past_actions = targets[:, :action_context_length]
-            target_actions = targets[:, action_context_length:]
+            # Normalize the target actions
+            joint_targets = normalizer.normalize(joint_targets)
 
+            # Reset the gradients
             optimizer.zero_grad()
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:112]: Why do we reset gradients?

File: ddlitlab2024/ml/training/train.py [hunk 3]
```diff
@@ -1,130 +1,129 @@
...
 
             # Sample a random timestep for each trajectory in the batch
             random_timesteps = (
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:115]: I'm confused as to what is random_timesteps here.
What dimensions does the tensor have? Is `joint_targets.size(0)` here 20, because of 20 degrees of freedom?

Additionally, why do we send the result of `randint` to device and not the whole `random_timesteps`?

Also we have a typing mismatch, because `random_timesteps` is not explicitly an `IntTensor`, which `add_noise` expects.

File: ddlitlab2024/ml/training/train.py [hunk 4]
```diff
@@ -1,130 +1,129 @@
...
-                torch.randint(0, scheduler.config["num_train_timesteps"], (batch_size,)).long().to(device)
+                torch.randint(0, scheduler.config["num_train_timesteps"], (joint_targets.size(0),)).long().to(device)
             )
 
             # Sample noise to add to the entire trajectory
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:119]: What kinda noise is it?

File: ddlitlab2024/ml/training/train.py [hunk 5]
```diff
@@ -137,8 +136,8 @@
             lr_scheduler.step()
             ema.update()
 
-            if (batch + 1) % 100 == 0:
-                print(f"Epoch {epoch}, Loss: {mean_loss / batch}, LR: {lr_scheduler.get_last_lr()[0]}")
+            if (i + 1) % 5 == 0:
+                print(f"Epoch {epoch}, Loss: {mean_loss / i}, LR: {lr_scheduler.get_last_lr()[0]}")
```
Comment (texhnolyze) [inline ddlitlab2024/ml/training/train.py:140]: How about using the logger or is there a reason no to?

**Review by texhnolyze — CHANGES_REQUESTED** [review]

File: ddlitlab2024/ml/training/train.py (hunk 2 above)
Comment (Flova) [reply ddlitlab2024/ml/training/train.py:112]: Because we used the gradients already and adjusted the weights using them. So they are no longer usefull.

File: ddlitlab2024/ml/training/train.py (hunk 4 above)
Comment (Flova) [reply ddlitlab2024/ml/training/train.py:119]: Gaussian iirc

File: ddlitlab2024/ml/training/train.py (hunk 5 above)
Comment (Flova) [reply ddlitlab2024/ml/training/train.py:140]: The code predates the logger, so we can integrate it in the future. Together with other things like wandb.

File: ddlitlab2024/ml/training/train.py [hunk 6]
```diff
@@ -1,130 +1,129 @@
...
+    logger.info("Create dataset objects")
+    dataset = DDLITLab2024Dataset()
+    num_workers = 5
+    dataloader = DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=True,
+        collate_fn=DDLITLab2024Dataset.collate_fn,
+        persistent_workers=num_workers > 1,
+        # prefetch_factor=10 * num_workers,
```
[the remark this answers is not in the document]
Comment (Flova) [reply ddlitlab2024/ml/training/train.py:46]: @jaagut what do you think?

[the remark this answers is not in the document]
Comment (texhnolyze) [reply ddlitlab2024/ml/training/train.py:57]: Will leave it in for testing.

**Review by texhnolyze — APPROVED** [review]
